src/esp32/board.rs

```diff
@@ -35,15 +39,119 @@ pub(crate) fn register_models(registry: &mut ComponentRegistry) {
     }
 }
 
+fn install_gpio_isr_service() -> anyhow::Result<()> {
+    static GPIO_ISR_SERVICE_INSTALLED: Lazy<Arc<OnceCell<()>>> =
+        Lazy::new(|| Arc::new(OnceCell::new()));
+    GPIO_ISR_SERVICE_INSTALLED.get_or_try_init(|| {
+        unsafe {
+            esp!(gpio_install_isr_service(ESP_INTR_FLAG_IRAM as i32))?;
+        };
+        Ok::<(), anyhow::Error>(())
+    })?;
+    Ok(())
+}
+
+/// GPIOPin is a wrapper for a pin on ESP32 as represented in esp-idf-hal
+/// and esp-idf-sys. This exists so that all micro-RDK drivers can interact
+/// with pins through the board instance and avoid conflicting uses of pins
+/// by multiple processes
+pub struct GPIOPin {
+    pin: i32,
+    driver: PinDriver<'static, AnyIOPin, InputOutput>,
+    pull: Pull,
+    interrupt_type: Option<InterruptType>,
+    event_count: Arc<AtomicU32>,
+}
+
+impl GPIOPin {
+    pub fn new(pin: i32, pull: Option<Pull>) -> anyhow::Result<Self> {
+        let mut driver = PinDriver::input_output(unsafe { AnyIOPin::new(pin) })?;
+        let pull = pull.unwrap_or(Pull::Up);
+        driver.set_pull(pull)?;
+        Ok(Self {
+            pin,
+            driver,
+            pull,
+            interrupt_type: None,
+            event_count: Arc::new(AtomicU32::new(0)),
+        })
+    }
+
+    pub fn pin(&self) -> i32 {
+        self.pin
+    }
+
+    pub fn is_high(&self) -> bool {
+        self.driver.is_high()
+    }
+
+    pub fn set_high(&mut self) -> anyhow::Result<()> {
+        self.driver
+            .set_high()
+            .map_err(|e| anyhow::anyhow!("couldn't set pin {} high {}", self.pin, e))
+    }
+
+    pub fn set_low(&mut self) -> anyhow::Result<()> {
+        self.driver
+            .set_low()
+            .map_err(|e| anyhow::anyhow!("couldn't set pin {} low {}", self.pin, e))
+    }
+
+    pub fn is_interrupt(&self) -> bool {
+        self.interrupt_type.is_some()
+    }
+
+    pub fn setup_interrupt(&mut self, intr_type: InterruptType) -> anyhow::Result<()> {
+        match &self.interrupt_type {
+            Some(existing_type) => {
+                if *existing_type == intr_type {
+                    return Ok(());
+                }
+            }
+            None => {
+                self.interrupt_type = Some(intr_type);
+            }
+        };
+        install_gpio_isr_service()?;
+        self.driver.set_interrupt_type(intr_type)?;
+        self.event_count.store(0, Ordering::Relaxed);
+        unsafe {
+            // we can't use the subscribe method on PinDriver to add the handler
+            // because it requires an FnMut with a static lifetime. A possible follow-up
```

Comment:
We should keep that in mind for when we are going to redesign the Board interface